custom-images/shell_script_generator.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2019 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#            http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Shell script based image creation workflow generator.
+"""
+
+from datetime import datetime
+
+
+_template = """
+#!/usr/bin/env bash
+
+# Script for creating Dataproc custom image.
+
+set -euxo pipefail
+
+disk_created="false"
+vm_created="false"
+
+function exit_handler() {{
+  echo 'Cleaning up before exiting.'
+
+  if [[ "$vm_created" == "true" ]]; then
+    echo 'Deleting VM instance.'
+    gcloud compute instances delete {image_name}-install \
+        --project={project_id} --zone={zone} -q
+  fi
+
+  if [[ "$disk_created" == "true" && "$vm_created" == "false" ]]; then
+    echo 'Deleting disk.'
+    gcloud compute disks delete {image_name}-install --project={project_id} --zone={zone} -q
+  fi
+
+  echo 'Uploading local logs to GCS bucket.'
+  gsutil -m rsync -r /tmp/{run_id}/logs gs://dagang-custom-images/{run_id}/logs/
+
+  sleep 5
```

Comment:
Why sleep here?

Reply:
No special reason, just want the user to notice it is exiting.

Reply:
How sleeps will help to notice exit? Maybe log some exit message?

Reply:
Done.